### PR TITLE
refactor: move reduction stack building out of parse_args.cc

### PR DESCRIPTION
**Pull request by lalo**
- most of the code from parse_args.cc moved to reduction_stack.cc
- build state is not part of all object anymore, see changes in global_data.h

### CODE REVIEW

File: vowpalwabbit/cached_learner.h
Comment (olgavrou) [inline vowpalwabbit/cached_learner.h:10]: should _cached be private here?

File: vowpalwabbit/reduction_stack.h
Comment (olgavrou) [inline vowpalwabbit/reduction_stack.h:20]: nit: `same signature as ...` do we want this as a comment in the code? maybe a comment in the PR for tracking?

File: vowpalwabbit/reduction_stack.h
Comment (olgavrou) [inline vowpalwabbit/reduction_stack.h:11]: `status_quo` -> `default_reduction_stack_setup` or something more descriptive, `status_quo` will loose it's meaning in a couple of months

File: vowpalwabbit/reduction_stack.h
Comment (olgavrou) [inline vowpalwabbit/reduction_stack.h:21]: I think this needs an `override` since it's implementing a pure virtual function from base

File: vowpalwabbit/cached_learner.h
Comment (lalo) [reply vowpalwabbit/cached_learner.h]: done

File: vowpalwabbit/reduction_stack.h
Comment (lalo) [reply vowpalwabbit/reduction_stack.h]: done

File: vowpalwabbit/reduction_stack.h
Comment (lalo) [reply vowpalwabbit/reduction_stack.h]: done

File: vowpalwabbit/reduction_stack.cc
Comment (lalo) [inline vowpalwabbit/reduction_stack.cc:220]: this also could be moved from all struct to default_reduction_stack_setup

**Review by olgavrou — APPROVED** [review]

**Comment by lalo** [comment]
This should go before https://github.com/VowpalWabbit/vowpal_wabbit/pull/2971

File: vowpalwabbit/reduction_stack.h
Comment (rajan-chari) [inline vowpalwabbit/reduction_stack.h:7]: setup_fn here is a pure virtual class with instantiations that create their own reduction stack.   Can the same thing can not be achieved if setup_fn decides what stack to instantiate?  This achieves the same thing without adding another parameter to the setup function of each reduction.

Comment (lalo) [reply vowpalwabbit/reduction_stack.h:7]: - The concern of adding an extra parameter is interesting, since in another PR after this one you could easily collapse options and all parameters inside setup_base_fn interface. If you care about parameter count, the count would get reduced to 1 and this 'object' starts to represent exactly what it is: a collection of configs and a magic box that gives you the base learner. You get the side benefit of having specific APIs that hide away VW all global state.
- Yes, you could achieve the same thing with a separate and alternate impls of current setup functions. I'm proposing that you can get for free with a single impl of setup function by breaking that coupling between setup functions all over the place and parse_args.h:setup_base fn. You can override this dependency on runtime by following the proposed interface in this PR.

Comment (lalo) [reply vowpalwabbit/reduction_stack.h:7]: https://github.com/VowpalWabbit/vowpal_wabbit/blob/d5bb16e2227148082aadf48daa6be050f139321d/vowpalwabbit/parse_args.cc#L1415

You can imagine alternate builders with other behaviour, like cached_learner that always returns the learner which it was constructed with like so: https://github.com/VowpalWabbit/vowpal_wabbit/blob/d5bb16e2227148082aadf48daa6be050f139321d/vowpalwabbit/cached_learner.h

Comment (rajan-chari) [reply vowpalwabbit/reduction_stack.h:7]: Makes sense.  setup_fn evolving to incorporate a factory for base reductions and other facilities that a reduction needs seems like a good general direction.  This makes it easier for a reduction to live in a different host outside vw.exe as long as the host provides all the same facilities.... like a factory for the next reduction, a random number generator, log facility, state save/resume facility etc.

**Review by rajan-chari — APPROVED** [review]

File: vowpalwabbit/binary.h
Comment (kumpera) [inline vowpalwabbit/binary.h:12]:
```suggestion
VW::LEARNER::base_learner* binary_setup(setup_base_fn& setup_base, VW::config::options_i& options, vw& all);
```
nit: arg name consistency.

File: vowpalwabbit/global_data.h
Comment (kumpera) [inline vowpalwabbit/global_data.h:281]: The learner_builder should be a pointer to an abstract object, otherwise we won't be able to effectively override its behavior.

File: vowpalwabbit/reduction_stack.h
Comment (kumpera) [inline vowpalwabbit/reduction_stack.h:13]: Public fields is usually bad.

**Review by kumpera — APPROVED** [review]
One thing about this PR that doesn't address is the decoupling between the transient state of build the reduction stack and the final state of the object.

It's an usual sign of problematic design when long lived objects retain vestigial setup data. In this case default_reduction_stack_setup keeps reduction_stack around.

I don't think this issue should block this PR as it's progress in the right direction.

File: vowpalwabbit/reduction_stack.h
Comment (lalo) [reply vowpalwabbit/reduction_stack.h:13]: I'll address this with https://github.com/VowpalWabbit/vowpal_wabbit/pull/3176

File: vowpalwabbit/parse_args.cc
Comment (lalo) [reply vowpalwabbit/parse_args.cc:1587]: This is a 'thing' I've seen across the code, audit_regressor being checked everytime we print something. Not sure why is that but it's big enough to get its own PR or refactor.

File: vowpalwabbit/global_data.h
Comment (lalo) [reply vowpalwabbit/global_data.h:281]: This will be removed after merging with https://github.com/VowpalWabbit/vowpal_wabbit/pull/3176

File: vowpalwabbit/reduction_stack.h
Comment (lalo) [reply vowpalwabbit/reduction_stack.h]: done

File: python/pylibvw.cc
Comment (lalo) [inline python/pylibvw.cc:20]:
```suggestion
```

File: vowpalwabbit/parse_args.cc
Comment (lalo) [inline vowpalwabbit/parse_args.cc:1419]: added this as a 'canary in the coalmine' for future PRs that expect learner_builder instance that is being passed thru setup reduction functions to be a long-lived one, when it should not be the case.

Comment (kumpera) [reply vowpalwabbit/parse_args.cc:1419]: Good idea.
We can revisit this later if we need some sort of clone & slice functionality for things like automl in case we want, for example, to have coin as a challenger to SGD. I currently have no idea on how that would it look like.

**Review by kumpera — APPROVED** [review]
👍

File: vowpalwabbit/global_data.h
Comment (lalo) [inline vowpalwabbit/global_data.h:58]:
```suggestion
```

**Pull request merged**
